### PR TITLE
Ensure userInfoFlow() emits null if offline and no data is cached

### PR DESCRIPTION
**Pull request by frett**
This change will cause userInfoFlow() to emit null if there is no data cached and fetching the data from the API fails for some reason.

The previous behavior would be the flow wouldn't emit anything if data wasn't cached and the user was offline. This led to consumers of this flow waiting indefinitely for something that may not happen.

### CODE REVIEW

File: gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt [hunk 1]
```diff
@@ -38,29 +40,33 @@ fun Credential.userInfoFlow() = flowOf(this).userInfoFlow()
...
         // emit UserInfo from the Credential tags
+        val cachedInfoFlow = cred.tagsFlow()
+            .map { it.cachedUserInfo }
+            .distinctUntilChanged()
         emitAll(
-            cred.tagsFlow()
-                .distinctUntilChanged()
-                .map { it.cachedUserInfo }
-                .dropWhile { it == null }
+            combine(cachedInfoFlow, loadAttempted) { data, loaded -> data to loaded }
```
[the remark this answers is not in the document]
Comment (frett) [reply gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt:63]: `to` is an inline function that is creating a data pair, so `loaded` is not an action here. `loaded` is the current value of the `loadAttempted` state holder.

File: gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt [hunk 2]
```diff
@@ -38,29 +40,33 @@ fun Credential.userInfoFlow() = flowOf(this).userInfoFlow()
...
                 }
 
                 delay((cred.tags.nextRefreshTime - System.currentTimeMillis()).coerceAtLeast(TimeConstants.MIN_IN_MS))
             }
         }
 
         // emit UserInfo from the Credential tags
+        val cachedInfoFlow = cred.tagsFlow()
+            .map { it.cachedUserInfo }
+            .distinctUntilChanged()
```
[the remark this answers is not in the document]
Comment (frett) [reply gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt:61]: It just changes what value is checked for distinctness.

by checking distinctness after we extract the `cachedUserInfo` it's only checking for changes in the data we care about and not the entire set of tags (which actually could have other changes not affecting our `cachedUserInfo`)

**Comment by frett** [comment]
downstream `null` is treated as there is currently no user info available for the current `Credential` object. This could mean they aren't authenticated or that it just isn't available.

This data flow is only for accessing userInfo, there is a separate `isAuthenticatedFlow` that should be utilized to determine authentication status.

**Review by wrandall22 — APPROVED** [review]

File: gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (frett) [reply gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/authfoundation/credential/Credential+Flow.kt]: @wrandall22 I made the data pair more obvious by switching to the `Pair()` constructor to create it. the `to` infix function is primarily used when creating maps of data like this:
```kotlin
mapOf(
  "a" to 1,
  "b" to 2,
)
```

**Pull request merged**